src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -0,0 +1,842 @@
+        }
+
+        private void WriteEndSlow(byte token)
+        {
+            Debug.Assert(_writerOptions.Indented || !_writerOptions.SkipValidation);
+
+            if (_writerOptions.Indented)
+            {
+                if (!_writerOptions.SkipValidation)
+                {
+                    ValidateEnd(token);
+                }
+                WriteEndIndented(token);
+            }
+            else
+            {
+                Debug.Assert(!_writerOptions.SkipValidation);
+                ValidateEnd(token);
+                WriteEndMinimized(token);
+            }
+        }
+
+        private void ValidateEnd(byte token)
+        {
+            if (_bitStack.CurrentDepth <= 0)
+                ThrowHelper.ThrowJsonWriterException(ExceptionResource.MismatchedObjectArray, token);
+
+            if (token == JsonConstants.CloseBracket)
+            {
+                if (_inObject)
+                {
+                    Debug.Assert(_tokenType != JsonTokenType.None);
+                    ThrowHelper.ThrowJsonWriterException(ExceptionResource.MismatchedObjectArray, token);
+                }
+            }
+            else
+            {
+                Debug.Assert(token == JsonConstants.CloseBrace);
+
+                if (!_inObject)
+                {
+                    ThrowHelper.ThrowJsonWriterException(ExceptionResource.MismatchedObjectArray, token);
+                }
+            }
+
+            _inObject = _bitStack.Pop();
+        }
+
+        private void WriteEndIndented(byte token)
+        {
+            // Do not format/indent empty JSON object/array.
+            if ((_tokenType == JsonTokenType.StartObject && token == JsonConstants.CloseBrace)
+                || (_tokenType == JsonTokenType.StartArray && token == JsonConstants.CloseBracket))
+            {
+                WriteEndMinimized(token);
+            }
+            else
+            {
+                int idx = 0;
+                WriteNewLine(ref idx);
+
+                int indent = Indentation;
+                while (true)
+                {
+                    bool result = JsonWriterHelper.TryWriteIndentation(_buffer.Slice(idx), indent, out int bytesWritten);
+                    idx += bytesWritten;
+                    if (result)
+                    {
+                        break;
+                    }
+                    indent -= bytesWritten;
+                    AdvanceAndGrow(idx);
+                    idx = 0;
+                }
+
+                while (_buffer.Length <= idx)
+                {
+                    AdvanceAndGrow(idx);
+                    idx = 0;
+                }
+                _buffer[idx++] = token;
+
+                Advance(idx);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNewLine(ref int idx)
+        {
+            // Write '\r\n' OR '\n', depending on OS
+            if (JsonWriterHelper.s_newLineUtf8.Length == 2)
```

Comment:
If we're not actually writing out s_newLineUtf8, why not just store a bool as to whether we should use \n or \r\n?